dotnet/src/InternalUtilities/connectors/AI/FunctionCallsProcessor.cs

```diff
@@ -0,0 +1,342 @@
+    /// <param name="chatMessageContent">The chat message content representing AI model response and containing function calls.</param>
+    /// <param name="chatHistory">The chat history to add function invocation results to.</param>
+    /// <param name="requestIndex">AI model function(s) call request sequence index.</param>
+    /// <param name="checkIfFunctionAdvertised">Callback to check if a function was advertised to AI model or not.</param>
+    /// <param name="kernel">The <see cref="Kernel"/>.</param>
+    /// <param name="resultSerializerOptions">Options to control function result serialization behavior.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests.</param>
+    /// <returns>Last chat history message if function invocation filter requested processing termination, otherwise null.</returns>
+    public async Task<ChatMessageContent?> ProcessFunctionCallsAsync(
+        ChatMessageContent chatMessageContent,
+        ChatHistory chatHistory,
+        int requestIndex,
+        Func<FunctionCallContent, bool> checkIfFunctionAdvertised,
+        Kernel? kernel,
+        JsonSerializerOptions? resultSerializerOptions,
+        CancellationToken cancellationToken)
+    {
+        var functionCalls = FunctionCallContent.GetFunctionCalls(chatMessageContent).ToList();
+
+        if (this._logger.IsEnabled(LogLevel.Debug))
+        {
+            this._logger.LogDebug("Function calls: {Calls}", functionCalls.Count);
+        }
+        if (this._logger.IsEnabled(LogLevel.Trace))
+        {
+            var messages = new List<string>(functionCalls.Count);
+            foreach (var call in functionCalls)
+            {
+                var argumentsString = call.Arguments is not null ? $"({string.Join(",", call.Arguments.Select(a => $"{a.Key}={a.Value}"))})" : "()";
+                messages.Add($"{call.FunctionName}{argumentsString}");
+            }
+            this._logger.LogTrace("Function calls: {Calls}", string.Join(", ", messages));
+        }
+
+        // Add the result message to the caller's chat history;
+        // this is required for AI model to understand the function results.
+        chatHistory.Add(chatMessageContent);
+
+        // We must send back a result for every function call, regardless of whether we successfully executed it or not.
+        // If we successfully execute it, we'll add the result. If we don't, we'll add an error.
+        for (int functionCallIndex = 0; functionCallIndex < functionCalls.Count; functionCallIndex++)
+        {
+            FunctionCallContent functionCall = functionCalls[functionCallIndex];
+
+            // Check if the function call has an exception.
+            if (functionCall.Exception is not null)
+            {
+                this.AddFunctionCallResultToChatHistory(chatHistory, functionCall, result: null, errorMessage: $"Error: Function call processing failed. {functionCall.Exception.Message}");
```

Comment:
This might introduce unnecessary behavioral change. Moreover, the function id is sent back to AI model, so adding the duplication information might not be necessary.

Reply:
Captured in this issue: https://github.com/microsoft/semantic-kernel/issues/8490, to be addressed later.